internal/backend/backend_retry.go

```diff
@@ -47,28 +46,9 @@ func (be *RetryBackend) retry(ctx context.Context, msg string, f func() error) e
 }
 
 // Save stores the data in the backend under the given handle.
-func (be *RetryBackend) Save(ctx context.Context, h restic.Handle, rd io.Reader) error {
-	seeker, ok := rd.(io.Seeker)
-	if !ok {
-		return errors.Errorf("reader %T is not a seeker", rd)
-	}
-
-	pos, err := seeker.Seek(0, io.SeekCurrent)
-	if err != nil {
-		return errors.Wrap(err, "Seek")
-	}
-
-	if pos != 0 {
-		return errors.Errorf("reader is not at the beginning (pos %v)", pos)
-	}
-
+func (be *RetryBackend) Save(ctx context.Context, h restic.Handle, length int, fn func() (io.Reader, error)) error {
 	return be.retry(ctx, fmt.Sprintf("Save(%v)", h), func() error {
-		_, err := seeker.Seek(0, io.SeekStart)
-		if err != nil {
-			return err
-		}
-
-		err = be.Backend.Save(ctx, h, rd)
+		err := be.Backend.Save(ctx, h, length, fn)
```

Comment:
shouldn't the old reader be closed here on retry?

Reply:
Which old reader?

Reply:
The API does not specify if `fn` should return a new reader or reset the same reader on retry. We either need to specify, and ideally enforce, that `fn` must always return the same reader, or assume it'll return new reader, in which case something will need to close the old one during retry.